### PR TITLE
Move the mutation rate from class to algorithm.

### PR DESCRIPTION
**Pull request by drvinceknight**
If there was a specific reason for having the mutation rate in the
params class
let me know but I've moved it to better reflect that it's a property of
the algorithm. This is more in line with the fact that the
representation we write to file does not hold the mutation rate for
example.

This doesn't stop having algorithms that would have different
mutation rates for different parameter instances (it's just that the
algorithm would have to keep track of it).

### CODE REVIEW

**Comment by Nikoleta-v3** [comment]
Looks great, just one comment.

In `utils.py` should the method `mutate` change to `mutate(self, mutation_rate)` in class `Params`?

**Comment by drvinceknight** [comment]
Thanks @Nikoleta-v3, good catch have fixed.

**Comment by marcharper** [comment]
> This doesn't stop having algorithms that would have different
mutation rates for different parameter instances (it's just that the
algorithm would have to keep track of it).

That was the point and this seemed like the easiest way to do it -- how would we do it now? With a map of  instances to mutation rates?

**Comment by drvinceknight** [comment]
That's the kind of thing I was thinking (it would need to keep track of what to do with births/deaths). How do you feel about having it as I've done it here and when/if we write an algorithm with different mutation rates, if a mapping type thing in the algorithm class doesn't do the trick then we cross that bridge?

**Comment by drvinceknight** [comment]
I might have a better idea (out walking, will do when I get back👍).

**Comment by drvinceknight** [comment]
> I might have a better idea (out walking, will do when I get back).

No my "better idea" doesn't feel right: I tried to have `def mutate(self)` but adding a `def receive_mutation_rate(self, mutation_rate)` that the algorithm would use but that feels like extra complications than necessary. I would suggest that when/if we algorithms with heterogeneous mutation rate we let the algorithm take care of it and simply use `def mutate(self, mutation_rate)`.

Another question:

```
88         if random.random() < self.mutation_rate / 10: 
```

I've only just noticed the `/10` there! Is there any reason for that? (unless I'm missing something trivial we need to remember to modify the values reported in the manuscripts).

**Comment by drvinceknight** [comment]
I've removed the `/10` in fc5ef4b.

**Comment by drvinceknight** [comment]
Also: any objection to renaming `mutation_rate -> mutation_probability`?

**Comment by marcharper** [comment]
> Also: any objection to renaming mutation_rate -> mutation_probability?

Sure.

>  How do you feel about having it as I've done it here and when/if we write an algorithm with different mutation rates, if a mapping type thing in the algorithm class doesn't do the trick then we cross that bridge?

For the speciation use case we discussed elsewhere my plan was to use the params as representations of the individuals and possibly allow the mutation rate to vary (to find the mutation-selection equilibrium, for example). So if we're going to move this inside the algorithm then I'd like to make sure we can handle that case. Otherwise I'd prefer to leave it as is and fix the non-logging of the mutation rate. We can also fix the logging in general -- the way I initially did it was quick and dirty, it would be better to write a dictionary, JSON, or something less ad hoc.

On the conceptual side, while I can certainly see why it seems better to have the mutation rate as a property of the evolutionary algorithm, I can think of cases in biological evolution where the mutation rate can vary for individuals of the same population, or between related species. I recall experimenting with a decaying mutation rate for the neural network strategy (since many training algorithms have the learning rate decrease over time), but I can't recall that it made a substantial difference.

**Comment by drvinceknight** [comment]
> For the speciation use case we discussed elsewhere my plan was to use the params as representations of the individuals and possibly allow the mutation rate to vary (to find the mutation-selection equilibrium, for example). So if we're going to move this inside the algorithm then I'd like to make sure we can handle that case.

I think my latest attempt ensures this can be handled now. The change is pretty minimilistic now:

- Change the argument passed from the algorithm to the instance to `kwargs` (as opposed to the `args` tuple) - this is just to make things a bit more robust going forward.
- The algorithm is passed a mutation rate and it passes this to the `kwargs`
- The algorithm just calls `instance.mutate()` as before

So the main effect is that a user doesn't need to create the `params_args` with anything but the arguments for the strategy itself (`num_states` etc...). I think this gives us flexibility going forward though with different approaches now being possible for different algorithms.

Let me know what you think :)

File: src/axelrod_dojo/archetypes/fsm.py [hunk 1]
```diff
@@ -83,10 +81,10 @@ def mutate_rows(rows, mutation_rate):
         return rows
 
     def mutate(self):
-        self.rows = self.mutate_rows(self.rows, self.mutation_rate)
-        if random.random() < self.mutation_rate / 10:
+        self.rows = self.mutate_rows(self.rows, self.mutation_probability)
```
Comment (marcharper) [inline src/axelrod_dojo/archetypes/fsm.py]: The scaling here was intentional so that some parts are mutated more often than others.

File: src/axelrod_dojo/archetypes/fsm.py [hunk 2]
```diff
@@ -83,10 +81,10 @@ def mutate_rows(rows, mutation_rate):
...
+        if random.random() < self.mutation_probability:
             self.initial_action = self.initial_action.flip()
-        if random.random() < self.mutation_rate / (10 * self.num_states):
```
Comment (marcharper) [inline src/axelrod_dojo/archetypes/fsm.py:89]: The scaling here was intentional so that some parts are mutated more often than others.

**Comment by marcharper** [comment]
Is this is going to break the other trainers? The name change to mutation_probability and args -> kwargs ...
This is better. Can we add a check so that if the agent params have mutation_probability then it's not overwritten, otherwise the algorithm will supply its "global" value? Then I think we're covered in all cases. Plus we can still just write the agent params dictionary to disk and they'll be correct in all cases.

File: src/axelrod_dojo/archetypes/fsm.py (hunk 2 above)
Comment (drvinceknight) [reply src/axelrod_dojo/archetypes/fsm.py:89]: Ok, I'm not a huge fan simply because of the ~~transference~~transparence in terms of how we need to describe the algorithm but I'll change it back. :+1:

**Comment by drvinceknight** [comment]
> Can we add a check so that if the agent params have mutation_probability then it's not overwritten, otherwise the algorithm will supply its "global" value? Then I think we're covered in all cases. Plus we can still just write the agent params dictionary to disk and they'll be correct in all cases.

Good shout will make that change.

I'll fix the other scripts. Perhaps we should/could remove them all together from here? As they're not tested any change etc becomes a bit of a pain.

**Comment by drvinceknight** [comment]
I believe b0e321b catches everything but let me know if not. (If you're happy could you `squash and merge` just to keep the history tidy and make the changelog super simple :+1:)

> Perhaps we should/could remove them all together from here? As they're not tested any change etc becomes a bit of a pain.

(Obviously we shouldn't do this until we've written the other parameter classes in to the library itself, so no rush at all with that call. Once we've moved them in to the library and they're all tested perhaps removing the scripts because of testing concerns won't be relevant.)

File: src/axelrod_dojo/archetypes/fsm.py (hunk 2 above)
Comment (marcharper) [reply src/axelrod_dojo/archetypes/fsm.py:89]: Ok, we can discuss it on another issue/PR but I found that changing the initial actions a lot wasn't useful.

Comment (drvinceknight) [reply src/axelrod_dojo/archetypes/fsm.py:89]: Yeah, not at all against the algorithmic choice (seems sensible), just that we need to make sure we describe it properly 👍:) NBD
